docs/misc/new_design.md

```diff
@@ -0,0 +1,169 @@
+
+```lean
+constant Point : Type
+```
+
+How about a line? Same.
+
+```lean
+constant Line : Type
+```
+
+How about the relations between points and lines?
+
+```lean
+constant lies_on : Point → Line → Prop
+constant between : Point → Point → Point → Prop  -- (between a b c) means "b is in between a and c"
+constant congruent {A : Type} : A → A → Prop
+constant distance : Point → Point → ℝ
+```
+
+How about axioms?
+
+```lean
+axiom distance_not_neg {p1 p2 : Point} : 0 ≤ distance p1 p2
+axiom distance_pos {p1 p2 : Point} : p1 ≠ p2 ↔ 0 < distance p1 p2
+```
+
+And this keeps on and on. Then we have a lot of lemmas and theorems and we still don't need to know what exactly a point is, we don't even need to know how to compute a distance.
+
+That's it. We don't need concrete types or computibility to reason about them. The semantic can end with their names and we don't need to know more underneath.
+
+This is the key insight one must have before formalizing a piece of mathematics.
+
+## Compatibility with mathlib
+
+A geometric product "contains" an inner product and an exterior product (or wedge product).
+
+They're established mathematic. There's already inner product space in mathlib and there would definitely be Grassmann Algebra in mathlib one day. And obviously we can't escape the general Clifford Algebra too.
+
+We must maintain compatibility with them, at least don't conflict with their existence.
+
+It's not a duplication. The development of mathlib is driven by professional mathematitians, they struggle for math generity but we're working at an abstraction level close to applications.
+
+We want to stand on such shoulders but we want good capsulation so that we don't want to worry about more abstract concepts and some too general cases. This would definitely leak, it might not be obvious in definitions but it would be very clear when you're proving something. For example, you have to deal with lattice and finsupp when proving things about linear algebra, and that should not be.
+
+## Operators
+
+We start with has_* that have absolutely no axioms with them, they don't have any properties and any behaviors. It's almost just a name. And we associate notations with them.
+
+Just like in https://github.com/leanprover-community/lean/blob/master/library/init/core.lean#L321 
+
+```lean
+class has_mul      (α : Type u) := (mul : α → α → α)
+```
+
+we could just:
+
+```lean
+class has_wedge (α : Type u) := (wedge : α → α → α)
+class has_inner (α : Type u) := (inner : α → α → α)
+```
+
+The latter presumably is already in mathlib and has a lot of structures and properties associated with it. We'll deal with that later.
+
+As for geometric product, I'm thinking about the following short name instead of `geomprod`, `gp` etc.:
```

Comment:
It turns out that `gmul` doesn't scale well, Scalar Product can't be `smul` because it's used by `has_scalar` which would be in our inheritance hierarchy.

Reply:
Can't we have `geometric_algebra.has_scalar_product.smul` to distinguish from `has_scalar.smul`? Or do namespaces not work well here?

Reply:
old structure is flat, so no matter where smul comes from, as long as it's in our inheritance hierarchy, it conflicts.

Better avoid it anyway.